emcc

```diff
@@ -1340,14 +1344,42 @@ try:
 
   js_transform_tempfiles = [final]
 
-  if memory_init_file:
+  if memory_init_file or shared.Settings.MEM_INIT_METHOD == 2:
     memfile = target + '.mem'
     shared.try_delete(memfile)
     def repl(m):
       # handle chunking of the memory initializer
       s = m.groups(0)[0]
       if len(s) == 0 and not shared.Settings.EMTERPRETIFY: return m.group(0) # emterpreter must have a mem init file; otherwise, don't emit 0-size ones
-      open(memfile, 'wb').write(''.join(map(lambda x: chr(int(x or '0')), s.split(','))))
+      membytes = [int(x or '0') for x in s.split(',')]
+      if not shared.Settings.EMTERPRETIFY:
+        while membytes and membytes[-1] == 0:
+          membytes.pop()
+        if not membytes:
+          return '';
+      if not memory_init_file:
+        crcTable = []
+        for i in range(256):
+          crc = i
+          for bit in range(8):
+            crc = (crc >> 1) ^ ((crc & 1) * 0xedb88320)
+          crcTable.append(crc)
+        crc = 0xffffffff
+        s = list(membytes)
+        n = len(s)
+        crc = crcTable[(crc ^ n) & 0xff] ^ (crc >> 8)
+        crc = crcTable[(crc ^ (n >> 8)) & 0xff] ^ (crc >> 8)
+        for i in s:
+          crc = crcTable[(crc ^ i) & 0xff] ^ (crc >> 8)
+        for i in range(4):
+          s.append((crc >> (8 * i)) & 0xff)
+        s = ''.join(map(chr, s))
+        s = s.replace('\\', '\\\\').replace("'", "\\'")
+        s = s.replace('\n', '\\n').replace('\r', '\\r')
+        def escape(x): return '\\x{:02x}'.format(ord(x.group()))
+        s = re.sub('[\x80-\xff]', escape, s)
+        return "var memoryInitializer = '%s';" % s
```

Comment:
we use chunking on the array, because (1) view source in browsers can fail on a massively long line, and (2) js engines would run out of stack trying to parse it.

we should investigate if a massively long string can be parsed by js engines (just by writing a little program that generates such a js program, without emscripten)